File: 1Urte desberdinetako materiala/Lorea 25-26/Android/Erronka-1/multi_intefaz/src/main/java/conexionbd/SocketClient.java
package conexionbd;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.scene.control.Alert;
import model.Genre;
import model.Writer;
import model.Liburu;
import model.IdazleData;
import java.io.*;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;

public class SocketClient {
    private static final String SERVER_HOST = "localhost";
    private static final int SERVER_PORT = 8080;
    private static final Gson gson = new Gson();
    
    public SocketClient() {
        // Constructor for compatibility
    }
    
    /**
     * Sends a request to the server and returns the response
     */
    private String sendRequest(String request) {
        try (Socket socket = new Socket(SERVER_HOST, SERVER_PORT);
             PrintWriter out = new PrintWriter(socket.getOutputStream(), true);
             BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
            
            // Send request
            out.println(request);
            
            // Read response
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = in.readLine()) != null) {
                response.append(line);
            }
            
            return response.toString();
            
        } catch (IOException e) {
            //Logger.getLogger(SocketClient.class.getName()).log(Level.SEVERE, "Error communicating with server", e);
            showAlert(Alert.AlertType.ERROR, "Connection Error", "Could not connect to server: " + e.getMessage());
            return null;
        }
    }
    
    /**
     * Get all authors from server
     */
    public ObservableList<Writer> getAuthors() {
        ObservableList<Writer> obs = FXCollections.observableArrayList();
        
        try {
            String request = "GET_AUTHORS";
            String response = sendRequest(request);
            
            if (response != null && !response.isEmpty()) {
                java.lang.reflect.Type type = new TypeToken<List<Writer>>(){}.getType();
                List<Writer> writers = gson.fromJson(response, type);
                obs.addAll(writers);
            }
            
        } catch (Exception ex) {
            //Logger.getLogger(Writer.class.getName()).log(Level.SEVERE, null, ex);
            System.out.println("Error al obtener los autores: " + ex.getMessage());
        }
        
        return obs;
    }
    
    /**
     * Get all genres from server
     */
    public ObservableList<Genre> getGenre() {
        ObservableList<Genre> obs = FXCollections.observableArrayList();
        
        try {
            String request = "GET_GENRES";
            String response = sendRequest(request);
            
            if (response != null && !response.isEmpty()) {
                java.lang.reflect.Type type = new TypeToken<List<Genre>>(){}.getType();
                List<Genre> genres = gson.fromJson(response, type);
                obs.addAll(genres);
            }
            
        } catch (Exception ex) {
            //Logger.getLogger(Writer.class.getName()).log(Level.SEVERE, null, ex);
            System.out.println("Error al obtener los generos: " + ex.getMessage());
        }
        
        return obs;
    }
    
    /**
     * Insert a new book
     */
    public void insertBook(String izenburua, int idIdazlea) {
        insertBook(izenburua, idIdazlea, new ArrayList<>());
    }
    
    /**
     * Insert a new book with genres
     */
    public void insertBook(String izenburua, int idIdazlea, List<String> generoak) {
        try {
            // Create the request with genres as a comma-separated string
            String genresString = String.join(",", generoak);
            String request = "INSERT_BOOK:" + izenburua + ":" + idIdazlea + ":" + genresString;
            String response = sendRequest(request);
            
            if (response != null && response.equals("SUCCESS")) {
                showAlert(Alert.AlertType.INFORMATION, "Success", "Book inserted successfully with genres: " + genresString);
            } else {
                showAlert(Alert.AlertType.ERROR, "Error", "Failed to insert book");
            }
            
        } catch (Exception e) {
            e.printStackTrace();
            showAlert(Alert.AlertType.ERROR, "Error", "Error inserting book: " + e.getMessage());
        }
    }
    
    /**
     * Insert a new writer and return the generated ID
     */
    public int insertWriter(String newWriter) {
        try {
            String request = "INSERT_WRITER:" + newWriter;
            String response = sendRequest(request);
            
            if (response != null && response.startsWith("SUCCESS:")) {
                String[] parts = response.split(":");
                if (parts.length > 1) {
                    return Integer.parseInt(parts[1]);
                }
            }
            
            showAlert(Alert.AlertType.ERROR, "Error", "Failed to insert writer");
            return -1;
            
        } catch (Exception e) {
            e.printStackTrace();
            showAlert(Alert.AlertType.ERROR, "Error", "Error inserting writer: " + e.getMessage());
            return -1;
        }
    }
    
    /**
     * Insert book-genre relationship
     */
    public void insertBookGen(String newBookGen) {
        try {
            String request = "INSERT_BOOK_GENRE:" + newBookGen;
            String response = sendRequest(request);
            
            if (response != null && response.equals("SUCCESS")) {
                showAlert(Alert.AlertType.INFORMATION, "Success", "Book-genre relationship inserted successfully");
            } else {
                showAlert(Alert.AlertType.ERROR, "Error", "Failed to insert book-genre relationship");
            }
            
        } catch (Exception e) {
            e.printStackTrace();
            showAlert(Alert.AlertType.ERROR, "Error", "Error inserting book-genre relationship: " + e.getMessage());
        }
    }
    
    /**
     * Insert a new genre and return the generated ID
     */
    public int insertGenre(String newGenre) {
        try {
            String request = "INSERT_GENRE:" + newGenre;
            String response = sendRequest(request);
            
            if (response != null && response.startsWith("SUCCESS:")) {
                String[] parts = response.split(":");
                if (parts.length > 1) {
                    return Integer.parseInt(parts[1]);
                }
            }
            
            showAlert(Alert.AlertType.ERROR, "Error", "Failed to insert genre");
            return -1;
            
        } catch (Exception e) {
            e.printStackTrace();
            showAlert(Alert.AlertType.ERROR, "Error", "Error inserting genre: " + e.getMessage());
            return -1;
        }
    }
    
    /**
     * Show alert dialog
     */
    public void showAlert(Alert.AlertType alertType, String title, String message) {
        Alert alert = new Alert(alertType);
        alert.setTitle(title);
        alert.setHeaderText(null);
        alert.setContentText(message);
        alert.show();
    }
    
    /**
     * Get all books from server
     */
    public ObservableList<Liburu> getBooks() {
        ObservableList<Liburu> obs = FXCollections.observableArrayList();
        
        try {
            String request = "GET_BOOKS";
            String response = sendRequest(request);
            
            if (response != null && !response.isEmpty()) {
                java.lang.reflect.Type type = new TypeToken<List<Liburu>>(){}.getType();
                List<Liburu> books = gson.fromJson(response, type);
                obs.addAll(books);
            }
            
        } catch (Exception ex) {
            System.out.println("Error al obtener los libros: " + ex.getMessage());
        }
        
        return obs;
    }
    
    /**
     * Get all author data (with their books) from server
     */
    public ObservableList<IdazleData> getAuthorData() {
        ObservableList<IdazleData> obs = FXCollections.observableArrayList();
        
        try {
            String request = "GET_AUTHOR_DATA";
            String response = sendRequest(request);
            
            if (response != null && !response.isEmpty()) {
                java.lang.reflect.Type type = new TypeToken<List<IdazleData>>(){}.getType();
                List<IdazleData> authorData = gson.fromJson(response, type);
                obs.addAll(authorData);
            }
            
        } catch (Exception ex) {
            System.out.println("Error al obtener los datos de autores: " + ex.getMessage());
        }
        
        return obs;
    }
    
    /**
     * Close connection (no-op for socket client, but kept for compatibility)
     */
    public void cerrarConexion() {
        // No need to close anything for socket client
        System.out.println("Socket connection closed");
    }
} 

File: PMD 25-26/Erronka/app/src/main/java/com/example/erronka/conexionBD/SocketClient.kt
package com.example.erronka.conexionBD
import android.content.Context
import android.os.Looper
import android.widget.Toast
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.example.erronka.model.Genre
import com.example.erronka.model.IdazleData
import com.example.erronka.model.Liburu
import com.example.erronka.model.Writer
import java.io.BufferedReader
import java.io.InputStreamReader

import java.io.PrintWriter
import java.net.Socket//Zerbitzari batekin sareko konexioa egiteko erabiltzen da.
import android.os.Handler
/*Android-en ezin dugu UI thread blokeatu:
Sareko eskaera → ez duzu Thread.sleep() erabili nahi → app-a geldituko litzateke
Fitxategiak irakurtzea edo datu-basea → denbora luzea hartu dezake
Coroutine-ek aukera ematen du:
Lan luzea background thread batean
Kodea lineala dirudien moduan idatzi
UI thread ez blokeatu*/
import kotlinx.coroutines.*//sareko lana background thread batean egitea.
import java.net.InetSocketAddress //Zerbitzariaren IP helbidea + portua definitzeko.

/**
 * Dispatchers.IO: fitxategiak, sarea, datu-basea
 * Dispatchers.Main: UI (Toast, TextView…)
 */
/**
 * SocketClient: Zerbitzariaren komunikazioa kudeatzen du.
 * Coroutines erabiliko ditugu UI thread-a ez blokeatzeko.
 * JSON erabiltzen du datuak trukatzeko Gson-ekin.
 */
class SocketClient(private val context: Context) {

    // Zerbitzariaren host-a eta portua
    private val SERVER_HOST = "10.0.2.2" // localhost// Android emulator -> host makina
    private val SERVER_PORT = 8080
    private val gson = Gson() // JSON bihurketa egiteko Gson objektua
    //Zerbitzariak  JSON formatuan bidaltzen ditu datuak.
    /**
     * Zerbitzarira eskaera bidali eta erantzuna lortu.
     * Coroutine-ean exekutatu behar da (Dispatchers.IO)
     */
    /*suspend funtzioak background thread batean exekutatzen dira
    coroutine batekin, baina kodea lineala dirudien moduan idazten da.
    Denbora luzezko lana egiteko (sarea)
    UI thread-a ez blokeatzeko*/
    private suspend fun sendRequest(request: String): String? {
        return try {
            // Socketa konexio timeout batekin ireki
            val socket = Socket()
            try {
                // Zerbitzarira konektatzen da, 2 segundoko timeout-arekin
                socket.connect(InetSocketAddress(SERVER_HOST, SERVER_PORT), 2000)
                //Eskaera testu moduan bidaltzen da
                val out = PrintWriter(socket.getOutputStream(), true)
                out.println(request)
                out.flush()
                //withContext(Dispatchers.IO) → sareko lana background thread-eanZerbitzariak bidalitako erantzuna irakurtzen du
                val reader = BufferedReader(InputStreamReader(socket.getInputStream()))
                reader.readLine()
            } finally {
                socket.close()
            }
        } catch (e: Exception) {
            // Zerbitzaria deskonektatuta dagoenean, Toast Main thread-ean erakutsi
            CoroutineScope(Dispatchers.Main).launch {
                Toast.makeText(context, "Zerbitzaria deskonektatuta dago", Toast.LENGTH_LONG).show()
            }
            null
        }
    }
    // ----------------- AUTHORS -----------------
    //withContext(Dispatchers.IO): sareko lana background thread-ean
    suspend fun getAuthors(): MutableList<Writer> = withContext(Dispatchers.IO) {
        val authors = mutableListOf<Writer>()
        val response = sendRequest("GET_AUTHORS")
        if (!response.isNullOrEmpty()) {
            //val type = object TypeToken<> -- Anonymous object bat sortzen da informazio generikoa gordetzeko
            //.type -- TypeToken-ek barruan gordetzen duen Type objektua ateratzen dugu
            // Kasu hontan Writer-en lista bat
            val type = object : TypeToken<List<Writer>>() {}.type
            //Gson-ek:
            //JSON irakurtzen du
            //Objektu berriak sortzen ditu (Writer)
            //Zerrenda bat bueltatzen du
            authors.addAll(gson.fromJson(response, type))
        }
        authors
    }
    // ----------------- GENRES -----------------
    suspend fun getGenres(): MutableList<Genre> = withContext(Dispatchers.IO) {
        val genres = mutableListOf<Genre>()
        val response = sendRequest("GET_GENRES")
        if (!response.isNullOrEmpty()) {
            val type = object : TypeToken<List<Genre>>() {}.type
            genres.addAll(gson.fromJson(response, type))
        }
        genres
    }
    // ----------------- BOOKS -----------------
    suspend fun getBooks(): MutableList<Liburu> = withContext(Dispatchers.IO) {
        val books = mutableListOf<Liburu>()
        val response = sendRequest("GET_BOOKS")
        if (!response.isNullOrEmpty()) {
            val type = object : TypeToken<List<Liburu>>() {}.type
            books.addAll(gson.fromJson(response, type))
        }
        books
    }

    // ----------------- AUTHOR DATA -----------------
    suspend fun getAuthorData(): MutableList<IdazleData> = withContext(Dispatchers.IO) {
        val data = mutableListOf<IdazleData>()
        val response = sendRequest("GET_AUTHOR_DATA")
        if (!response.isNullOrEmpty()) {
            val type = object : TypeToken<List<IdazleData>>() {}.type
            data.addAll(gson.fromJson(response, type))
        }
        data
    }

    // ----------------- INSERT WRITER -----------------
    /**
     * insertWriter()
     *  1.-IO thread-ean exekutatu
     *  2.-Zerbitzarira mezua bidali
     *  3.-Erantzuna aztertu
     *      3.1.-SUCCESS → ID bueltatu
     *      3.2.-ERROR → Toast + -1
     */
    //newWriter: String -- Gehitu nahi dugun idazlearen izena
    //  Int = withContext(Dispatchers.IO)
    // Coroutine baten barruan thread-a aldatzeko erabiltzen da eta
    // bertan sortutako emaitza hemen bultatzeko
    //  Int bueltatzen du -- idazlearen ID-a baina

    suspend fun insertWriter(newWriter: String): Int = withContext(Dispatchers.IO) {
        /**
         * sendRequest(...)
         * Socket-a irekitzen du
         * Mezua zerbitzarira bidaltzen du
         * Erantzuna jasotzen du (String moduan)
         */
        val response = sendRequest("INSERT_WRITER:$newWriter")
        /**
         * erantzuna:
         * "SUCCESS:5": ondo (ID = 5)
         * null: errorea edo konexiorik ez
         */
        if (response != null && response.startsWith("SUCCESS:")) {
            /**
             * "SUCCESS:5" bada erantzuna
             * 5 itzuliko da hau da metodoak ID zenbakia bueltatzen du
             */
            response.split(":")[1].toInt()
        } else {
            withContext(Dispatchers.Main) {
                Toast.makeText(context, "Idazlea txertatzea huts egin du", Toast.LENGTH_LONG).show()
            }
            -1
        }
    }

    // ----------------- INSERT BOOK -----------------
    suspend fun insertBook(izenburua: String, authorId: Int, generoak: List<String> = emptyList()) = withContext(Dispatchers.IO) {
        val genresString = generoak.joinToString(",")
        val response = sendRequest("INSERT_BOOK:$izenburua:$authorId:$genresString")
        withContext(Dispatchers.Main) {
            if (response == "SUCCESS") {
                Toast.makeText(context, "Liburua txertatu da generoekin: $genresString", Toast.LENGTH_LONG).show()
            } else {
                Toast.makeText(context, "Liburua txertatzea huts egin du", Toast.LENGTH_LONG).show()
            }
        }
    }

    // ----------------- INSERT GENRE -----------------
    suspend fun insertGenre(newGenre: String): Int = withContext(Dispatchers.IO) {
        val response = sendRequest("INSERT_GENRE:$newGenre")
        if (response != null && response.startsWith("SUCCESS:")) {
            response.split(":")[1].toInt()
        } else {
            withContext(Dispatchers.Main) {
                Toast.makeText(context, "Genero txertatzea huts egin du", Toast.LENGTH_LONG).show()
            }
            -1
        }
    }

    // ----------------- INSERT BOOK-GENRE -----------------
    suspend fun insertBookGen(newBookGen: String) = withContext(Dispatchers.IO) {
        val response = sendRequest("INSERT_BOOK_GENRE:$newBookGen")
        withContext(Dispatchers.Main) {
            if (response == "SUCCESS") {
                Toast.makeText(context, "Liburu-genero erlazioa txertatu da", Toast.LENGTH_LONG).show()
            } else {
                Toast.makeText(context, "Liburu-genero erlazioa txertatzea huts egin du", Toast.LENGTH_LONG).show()
            }
        }
    }
    /**
     * Android Toast erakutsi
     */
    fun showToast(message: String) {
        CoroutineScope(Dispatchers.Main).launch {
            Toast.makeText(context, message, Toast.LENGTH_LONG).show()
        }
    }

    /**
     * Konexioa ixteko metodoa (socketa berez ixtea gertatzen da)
     */
    fun closeConnection() {
        showToast("Socket konexioa itxi da.")
    }
}
